### PR TITLE
fix: enforce kurtosis.yml validations in import_module and read_file; package name inside kurtosis.yml must be valid and is same as the path where kurtosis.yml exists

### PR DESCRIPTION
**Pull request by Peeeekay**
…d in the same location as the package name inside kurtosis.yml

Changelog picked up from commits here:

fix: enforce kurtosis.yml validations in import_module and read_file; package name inside kurtosis.yml must be valid and is same as the path where kurtosis.yml exists

### CODE REVIEW

File: core/server/api_container/server/startosis_engine/startosis_packages/git_package_content_provider/git_package_content_provider.go [hunk 1]
```diff
@@ -293,6 +304,31 @@ func getReferenceName(repo *git.Repository, parsedURL *ParsedGitURL) (plumbing.R
 	return "", false, nil
 }
 
+func validateKurtosisPackage(kurtosisYaml *yaml_parser.KurtosisYaml, absPathToPackageWithKurtosisYml string, packageDir string) *startosis_errors.InterpretationError {
+	// first check whether the package name extracted from yaml is valid
+	// re-using parseGitURL as it already does some validations
+	packageName := kurtosisYaml.GetPackageName()
+	if strings.HasSuffix(packageName, string(os.PathSeparator)) {
```
Comment (Peeeekay) [inline core/server/api_container/server/startosis_engine/startosis_packages/git_package_content_provider/git_package_content_provider.go]: As per the docs, the schema is github.com/author/repo ---> so `github.com/author/repo/`  is not valid.

**Comment by Peeeekay** [comment]
This PR is extension to: https://github.com/kurtosis-tech/kurtosis/pull/16; in this the idea is to validate kurtosis.yml during `import_module` and `read_file` i.e does the locator follows the schema and is the package name same as the locator 

Say kurtosis.yml is in this path - `github.com/tech/test/package/kurtosis.yml` --> the package name inside the yml must be
```
name: github.com/tech/test/package
```

File: core/server/api_container/server/startosis_engine/startosis_packages/git_package_content_provider/git_package_content_provider.go [hunk 2]
```diff
@@ -293,6 +304,31 @@ func getReferenceName(repo *git.Repository, parsedURL *ParsedGitURL) (plumbing.R
...
+		return startosis_errors.NewInterpretationError("Error occurred while validating package name, found '%v' and expected '%v' without trailing '%v'", packageName, packageNameWithoutSuffix, string(os.PathSeparator))
+	}
+
+	_, err := parseGitURL(packageName)
+	if err != nil {
+		return startosis_errors.WrapWithInterpretationError(err, "Error occurred while validating package name: %v which is found here: '%v'", kurtosisYaml.GetPackageName(), absPathToPackageWithKurtosisYml)
+	}
+
+	// get package name from absolute path to package
+	packageNameFromPackagePath := strings.Replace(absPathToPackageWithKurtosisYml, packageDir, startosis_constants.GithubPrefix, 1)
```
Comment (h4ck3rk3y) [inline core/server/api_container/server/startosis_engine/startosis_packages/git_package_content_provider/git_package_content_provider.go]: lets replace this magic constant "1"

File: core/server/api_container/server/startosis_engine/startosis_packages/git_package_content_provider/git_package_content_provider.go [hunk 3]
```diff
@@ -293,6 +304,31 @@ func getReferenceName(repo *git.Repository, parsedURL *ParsedGitURL) (plumbing.R
...
+	removeKurtosisYmlFromPackageName := strings.TrimSuffix(packageNameFromPackagePath, fmt.Sprintf("%v%v", string(os.PathSeparator), startosis_constants.KurtosisYamlName))
```
Comment (h4ck3rk3y) [inline core/server/api_container/server/startosis_engine/startosis_packages/git_package_content_provider/git_package_content_provider.go]: Can we  do path.base or something?

File: core/server/commons/yaml_parser/kurtosis_yaml_test.go [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+	require.NotNil(t, err)
+	require.ErrorContains(t, err, fmt.Sprintf("Error occurred while reading the contents of %v", path))
+}
+
+func Test_parseKurtosisYamlInternal_IncorrectYaml(t *testing.T) {
+	mockRead := func(filename string) ([]byte, error) {
+		return sampleInCorrectYaml, nil
+	}
+
+	path := "/root/kurtosis.yml"
```
Comment (h4ck3rk3y) [inline core/server/commons/yaml_parser/kurtosis_yaml_test.go]: lets make this a constant

File: core/server/commons/yaml_parser/kurtosis_yaml_test.go [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+	_, err := parseKurtosisYamlInternal(path, mockRead)
+	require.Nil(t, err)
+	require.Equal(t, "", "")
```
Comment (h4ck3rk3y) [inline core/server/commons/yaml_parser/kurtosis_yaml_test.go]: whats going on here?

File: core/server/api_container/server/startosis_engine/startosis_packages/git_package_content_provider/git_package_content_provider.go [hunk 6]
```diff
@@ -293,6 +304,31 @@ func getReferenceName(repo *git.Repository, parsedURL *ParsedGitURL) (plumbing.R
 	return "", false, nil
 }
 
+func validateKurtosisPackage(kurtosisYaml *yaml_parser.KurtosisYaml, absPathToPackageWithKurtosisYml string, packageDir string) *startosis_errors.InterpretationError {
+	// first check whether the package name extracted from yaml is valid
```
Comment (h4ck3rk3y) [inline core/server/api_container/server/startosis_engine/startosis_packages/git_package_content_provider/git_package_content_provider.go]: does this comment apply to line 316?

File: core/server/api_container/server/startosis_engine/startosis_packages/git_package_content_provider/git_package_content_provider.go [hunk 7]
```diff
@@ -293,6 +304,31 @@ func getReferenceName(repo *git.Repository, parsedURL *ParsedGitURL) (plumbing.R
...
+	// re-using parseGitURL as it already does some validations
+	packageName := kurtosisYaml.GetPackageName()
+	if strings.HasSuffix(packageName, string(os.PathSeparator)) {
+		packageNameWithoutSuffix := strings.TrimSuffix(packageName, string(os.PathSeparator))
+		return startosis_errors.NewInterpretationError("Error occurred while validating package name, found '%v' and expected '%v' without trailing '%v'", packageName, packageNameWithoutSuffix, string(os.PathSeparator))
```
Comment (h4ck3rk3y) [inline core/server/api_container/server/startosis_engine/startosis_packages/git_package_content_provider/git_package_content_provider.go]: Can we simply say trailing "/" isn't valid in package names?

Comment (h4ck3rk3y) [reply core/server/api_container/server/startosis_engine/startosis_packages/git_package_content_provider/git_package_content_provider.go]: i am concerned that the package name initially would be

free/cofee/

you'd tell them that you expected free/coffee and then complain with a separate error when they try that, can be confusing

File: core/server/api_container/server/startosis_engine/startosis_packages/git_package_content_provider/git_package_content_provider.go [hunk 8]
```diff
@@ -293,6 +304,31 @@ func getReferenceName(repo *git.Repository, parsedURL *ParsedGitURL) (plumbing.R
 	return "", false, nil
 }
 
+func validateKurtosisPackage(kurtosisYaml *yaml_parser.KurtosisYaml, absPathToPackageWithKurtosisYml string, packageDir string) *startosis_errors.InterpretationError {
+	// first check whether the package name extracted from yaml is valid
+	// re-using parseGitURL as it already does some validations
+	packageName := kurtosisYaml.GetPackageName()
+	if strings.HasSuffix(packageName, string(os.PathSeparator)) {
+		packageNameWithoutSuffix := strings.TrimSuffix(packageName, string(os.PathSeparator))
```
Comment (h4ck3rk3y) [inline core/server/api_container/server/startosis_engine/startosis_packages/git_package_content_provider/git_package_content_provider.go]: can string(os.PathSeparator) be a constant, we use it at a few places

File: core/server/api_container/server/startosis_engine/startosis_constants/startosis_constants.go [hunk 9]
```diff
@@ -3,8 +3,8 @@ package startosis_constants
 const (
 	MainFileName     = "main.star"
 	KurtosisYamlName = "kurtosis.yml"
-
-	EmptyInputArgs = "{}" // empty JSON
+	GithubPrefix     = "github.com"
```
Comment (h4ck3rk3y) [inline core/server/api_container/server/startosis_engine/startosis_constants/startosis_constants.go]: do we use this at multiple places?

Comment (h4ck3rk3y) [reply core/server/api_container/server/startosis_engine/startosis_constants/startosis_constants.go]: parse_git_url could use this

File: core/server/api_container/server/startosis_engine/startosis_packages/git_package_content_provider/git_package_content_provider.go [hunk 10]
```diff
@@ -293,6 +304,31 @@ func getReferenceName(repo *git.Repository, parsedURL *ParsedGitURL) (plumbing.R
...
+		return startosis_errors.NewInterpretationError("Error occurred while validating package name, found '%v' and expected '%v' without trailing '%v'", packageName, packageNameWithoutSuffix, string(os.PathSeparator))
+	}
+
+	_, err := parseGitURL(packageName)
+	if err != nil {
+		return startosis_errors.WrapWithInterpretationError(err, "Error occurred while validating package name: %v which is found here: '%v'", kurtosisYaml.GetPackageName(), absPathToPackageWithKurtosisYml)
```
Comment (h4ck3rk3y) [inline core/server/api_container/server/startosis_engine/startosis_packages/git_package_content_provider/git_package_content_provider.go]: nit: s/which is found here/in Kurtosis.yml at '%v'/

File: core/server/commons/yaml_parser/kurtosis_yaml_test.go [hunk 11]
```diff
@@ -0,0 +1,44 @@
+package yaml_parser
+
+import (
+	"fmt"
+	"github.com/stretchr/testify/require"
+	"io"
+	"testing"
+)
+
+var sampleCorrectYaml = []byte(`name: github.com/test/test`)
```
Comment (h4ck3rk3y) [inline core/server/commons/yaml_parser/kurtosis_yaml_test.go]: is this a valid Kurtosis.yml

```
name: github.com/test/test
description: this shouldn't be valid?
```

**Review by h4ck3rk3y — APPROVED** [review]

File: core/server/commons/yaml_parser/kurtosis_yaml_test.go (hunk 11 above)
Comment (Peeeekay) [reply core/server/commons/yaml_parser/kurtosis_yaml_test.go]: yeah `test` is repo and `test` is author as well; it is. will change the name though will clarity.

**Comment by leeederek** [comment]
Hey @Peeeekay how is this different from #16

**Comment by Peeeekay** [comment]
@leeederek - this is internal change - just validates that the package name in `kurtosis.yml` matches the location where it is present

**Pull request merged**